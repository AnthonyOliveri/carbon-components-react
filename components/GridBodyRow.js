import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

export default function GridBodyRow(props) {
  const cells = props.columns.map(column => {
    const width = `${column.width}%`;
    const columnClass = classNames(
      'bx--grid-body-row__cell',
      column.overflow ? 'bx--grid-body-row__cell--overflow' : ''
    );

    return (
      <div className={columnClass} key={column.title} style={{ width }}>
        {column.render(props.data, props.index)}
      </div>
    );
  });

  const classes = ['bx--grid-body-row', props.className];

  let rowExpansion = '';
  if (props.rowExpanded) {
    rowExpansion = (
      <div className="bx--grid-body-row__row--expansion">
        {props.renderRowExpansion(props.data, props.index)}
      </div>
    );
  }

  return (
    <div className={classes.join(' ')}>
      <div className="bx--grid-body-row__row">
        {cells}
      </div>
      {rowExpansion}
    </div>
  );
}

GridBodyRow.defaultProps = {
  className: '',
  columns: [],
  data: {},
  renderRowExpansion: () => {},
  rowExpanded: false,
};

GridBodyRow.propTypes = {
  className: PropTypes.string,
  columns: PropTypes.array.isRequired,
  data: PropTypes.object.isRequired,
  index: PropTypes.number.isRequired,
  renderRowExpansion: PropTypes.func,
  rowExpanded: PropTypes.bool,
};
