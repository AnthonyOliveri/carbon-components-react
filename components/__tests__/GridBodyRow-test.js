import GridBodyRow from '../GridBodyRow';
import React from 'react';
import { shallow } from 'enzyme';

jest.unmock('../GridBodyRow');

const shallowRender = props => shallow(<GridBodyRow {...props} />);

describe('GridBodyRow', () => {
  describe('check if props are rendered correctly', () => {
    describe('standard row - no row expansion', () => {
      const columns = [
        {
          name: 'column1',
          title: 'column1',
          render: data => data.column1,
          width: 50,
          overflow: true,
        },
        {
          name: 'column2',
          title: 'column2',
          render: data => data.column2,
          width: 25,
        },
        {
          name: 'column3',
          title: 'column3',
          render: data => data.column3,
          width: 25,
        },
      ];
      const wrapper = shallowRender({
        className: 'testRow',
        columns,
        data: {
          column1: 'stuff',
          column2: 'more',
          column3: 'test',
        },
        index: 2,
        renderRowExpansion: null,
        rowExpanded: false,
      });

      it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe('standard row - row expanded', () => {
      const columns = [
        {
          name: 'column1',
          title: 'column1',
          render: data => data.column1,
          width: 50,
        },
        {
          name: 'column2',
          title: 'column2',
          render: data => data.column2,
          width: 25,
        },
        {
          name: 'column3',
          title: 'column3',
          render: data => data.column3,
          width: 25,
        },
      ];
      const data = {
        column1: 'stuff',
        column2: 'more',
        column3: 'test',
        expander: 'expandData',
      };
      const wrapper = shallowRender({
        className: 'testRow',
        columns,
        data,
        index: 2,
        renderRowExpansion: (rowData, index) => `${rowData.expander}-${index}`,
        rowExpanded: true,
      });

      it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
