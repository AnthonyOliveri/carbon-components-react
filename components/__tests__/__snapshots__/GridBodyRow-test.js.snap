exports[`GridBodyRow check if props are rendered correctly standard row - no row expansion renders correctly 1`] = `
<div
  className="bx--grid-body-row testRow">
  <div
    className="bx--grid-body-row__row">
    <div
      className="bx--grid-body-row__cell bx--grid-body-row__cell--overflow"
      style={
        Object {
          "width": "50%",
        }
      }>
      stuff
    </div>
    <div
      className="bx--grid-body-row__cell"
      style={
        Object {
          "width": "25%",
        }
      }>
      more
    </div>
    <div
      className="bx--grid-body-row__cell"
      style={
        Object {
          "width": "25%",
        }
      }>
      test
    </div>
  </div>
</div>
`;

exports[`GridBodyRow check if props are rendered correctly standard row - row expanded renders correctly 1`] = `
<div
  className="bx--grid-body-row testRow">
  <div
    className="bx--grid-body-row__row">
    <div
      className="bx--grid-body-row__cell"
      style={
        Object {
          "width": "50%",
        }
      }>
      stuff
    </div>
    <div
      className="bx--grid-body-row__cell"
      style={
        Object {
          "width": "25%",
        }
      }>
      more
    </div>
    <div
      className="bx--grid-body-row__cell"
      style={
        Object {
          "width": "25%",
        }
      }>
      test
    </div>
  </div>
  <div
    className="bx--grid-body-row__row--expansion">
    expandData-2
  </div>
</div>
`;
